compiler/rustc_mir_transform/src/check_unsafety.rs

```diff
@@ -387,41 +400,109 @@ pub(crate) fn provide(providers: &mut Providers) {
 
         if let hir::BlockCheckMode::UnsafeBlock(hir::UnsafeSource::UserProvided) = block.rules {
-            self.unsafe_blocks.push((block.hir_id, self.used_unsafe.contains(&block.hir_id)));
+            let used = match self.tcx.lint_level_at_node(UNUSED_UNSAFE, block.hir_id) {
+                (Level::Allow, _) => Some(SomeDisallowedInUnsafeFn),
+                _ => self.used_unsafe_blocks.get(&block.hir_id).copied(),
+            };
+            // only pushes if the contained `match` doesn't `return` early
+            self.unused_unsafe.push((
+                block.hir_id,
+                match (self.context, used) {
+                    (_, None) => UnusedUnsafe::Unused,
+                    (Context::Safe, Some(_))
+                    | (Context::UnsafeFn(_), Some(SomeDisallowedInUnsafeFn)) => {
+                        let previous_context = self.context;
+                        self.context = Context::UnsafeBlock(block.hir_id);
+                        intravisit::walk_block(self, block);
+                        self.context = previous_context;
+                        return;
+                    }
+                    (Context::UnsafeFn(hir_id), Some(AllAllowedInUnsafeFn(lint_root))) => {
+                        UnusedUnsafe::InUnsafeFn(hir_id, lint_root)
+                    }
+                    (Context::UnsafeBlock(hir_id), Some(_)) => UnusedUnsafe::InUnsafeBlock(hir_id),
+                },
+            ));
+        }
+        intravisit::walk_block(self, block);
+    }
+
+    fn visit_fn(
+        &mut self,
+        fk: intravisit::FnKind<'tcx>,
+        _fd: &'tcx hir::FnDecl<'tcx>,
+        b: hir::BodyId,
+        _s: rustc_span::Span,
+        _id: HirId,
+    ) {
+        if matches!(fk, intravisit::FnKind::Closure) {
+            self.visit_body(self.tcx.hir().body(b))
         }
     }
 }
 
 fn check_unused_unsafe(
     tcx: TyCtxt<'_>,
     def_id: LocalDefId,
-    used_unsafe: &FxHashSet<hir::HirId>,
-    unsafe_blocks: &mut Vec<(hir::HirId, bool)>,
-) {
-    let body_id = tcx.hir().maybe_body_owned_by(tcx.hir().local_def_id_to_hir_id(def_id));
+    used_unsafe_blocks: &FxHashMap<HirId, UsedUnsafeBlockData>,
+) -> Vec<(HirId, UnusedUnsafe)> {
+    let hir_id = tcx.hir().local_def_id_to_hir_id(def_id);
+    let body_id = tcx.hir().maybe_body_owned_by(hir_id);
 
     let body_id = match body_id {
         Some(body) => body,
         None => {
             debug!("check_unused_unsafe({:?}) - no body found", def_id);
-            return;
+            return vec![];
         }
     };
     let body = tcx.hir().body(body_id);
-    debug!("check_unused_unsafe({:?}, body={:?}, used_unsafe={:?})", def_id, body, used_unsafe);
 
-    let mut visitor = UnusedUnsafeVisitor { used_unsafe, unsafe_blocks };
+    let context = match tcx.hir().find(hir_id) {
+        Some(Node::Item(&hir::Item { kind: hir::ItemKind::Fn(ref sig, _, _), .. }))
+            if sig.header.unsafety == hir::Unsafety::Unsafe =>
```

Comment:
Oh, looking into what `fn_sig_by_hir_id`’s implementation does, I do realize that different item-like things aren’t even handled at all currently. The code is adapted from the logic previously present in the `fn is_enclosed` function that used to exist. Turns out, I never tested code examples such as

```rs
pub struct S;

impl S {
    pub unsafe fn f() {
        unsafe {
            unsf();
        }
    }
}

unsafe fn unsf() {}
```

This code doesn’t work properly on `nightly` (it doesn’t point to the containing `unsafe fn`) and doesn’t work even worse with the code currently in this PR (it doesn’t even report any unused `unsafe`ty at all…).

I suppose using `fn_sig_by_hir_id` should fix that nicely, I’ll also add test cases.

Reply:
I wouldn’t understand how `let-else` applies here.

Reply:
Fixed now.

Reply:
> I wouldn’t understand how `let-else` applies here.

I was thinking of something like:

```rust
let Some(sig) = fn_sig_by_hir_id(hir_id) else { bug!("cannot obtain signature item with body") };
let context = if sig.header.unsafety == hir::Unsafety::Unsafe { ... } else { ... };
```

It seems somewhat dangerous to default to `Context::Safe` when the function signature is not available.

Reply:
Dang it, I just lost a previous version of this response due to my desktop environment crashing… let me re-write the main points as far as I remember.

The `hir_id` here can refer to any body owner, not just functions; this includes closures… well actually, closures are ruled-out because they’re handled differently in `fn unsafety_check_result`… and constants, e.g. const items, constants in types, the length in an in array expression, etc. Hence, failure to get a `sig` is not a bug. Failure to `.find()` the id in the first place is already handled by `fn_sig_by_hir_id` internally with a call to `bug!`.

<hr>

Note that an incorrect `Context::Safe` is not actually dangerous. This context is only used for the traversal for `unused_unsafe` warnings. Worst thing that could happen is that an unnecessary unsafe block (e.g. one that’s appearing inside of an `unsafe fn`) is not reported as such.